source/mir/random/discrete.d

```diff
@@ -11,6 +11,212 @@ module mir.random.discrete;
 import std.traits : isNumeric;
 
 /**
+Setup a discrete distribution sampler.
+
+Params:
+    probs = probabilities of the individual, discrete values
+
+Returns:
+    A $(LREF Discrete) sampler that can be called to sample from the distribution.
+*/
+Discrete!T discrete(T)(const(T)[] probs)
+{
+    return Discrete!T(probs);
+}
+
+///
+unittest
+{
+    auto probs = [0.1, 0.2, 0.5, 0.2];
+    auto ds = discrete(probs);
+
+    // sample from the discrete distribution
+    auto obs = new uint[probs.length];
+    foreach (i; 0..10_000)
+        obs[ds()]++;
+}
+
+/**
+_Discrete distribution sampler that draws random values from a _discrete
+distribution given an array of the respective probabilities.
+
+Complexity: O(1) per sampling and O(n) for the initial setup, where n is the number
+            of discrete values.
+
+References:
+    Vose, Michael D. "A linear algorithm for generating random numbers with a given distribution."
+    IEEE Transactions on software engineering 17.9 (1991): 972-975.
+*/
+struct Discrete(T)
+    if (isNumeric!T)
+{
+
+    /// array with the original column value for a discrete value an
+    struct AltPair
+    {
+        T prob; /// probability p to select it by a coin toss, if this column is randomly picked
+        size_t alt; /// alternative value if coin toss at j fails
+    }
+
+    private AltPair[] arr;
+
+    /**
+    Initialize a discrete distribution sampler
+    Params:
+        probs = probabilities of the individual, discrete values
+
+    Complexity: O(n), where n is the number of discrete values
+    */
+    this(const(T)[] probs)
+    {
+        debug
+        {
+            import std.algorithm.iteration : sum;
+            import std.math : approxEqual;
+            assert(probs.sum.approxEqual(1.0), "Sum of the probabilities must be 1");
+        }
+        initialize(probs);
```

Comment:
As many other random algorithms also have a construction phase, it would be quite nice if we agree that the initialization is part of the struct initialization